storage/aptosdb/src/db_debugger/truncation/mod.rs

```diff
@@ -0,0 +1,342 @@
+        let state_merkle_db_version =
+            Self::get_current_version_in_state_merkle_db(&state_merkle_db)?
+                .expect("Current version of state merkle db must exist.");
+
+        assert!(self.target_version <= ledger_db_version);
+        assert!(ledger_db_version >= state_merkle_db_version);
+
+        let state_merkle_target_version =
+            Self::find_tree_root_at_or_before(&ledger_db, &state_merkle_db, self.target_version)?
+                .expect(&format!(
+                    "Could not find a valid root before or at version {}, maybe it was pruned?",
+                    self.target_version
+                ));
+
+        Self::truncate_state_merkle_db(&state_merkle_db, state_merkle_target_version)?;
+
+        let ledger_db = Arc::new(ledger_db);
+        Self::truncate_ledger_db(
+            Arc::clone(&ledger_db),
+            ledger_db_version,
+            self.target_version,
+            self.ledger_db_batch_size,
+        )?;
+
+        if state_merkle_target_version < self.target_version {
+            StateStore::catch_up_state_merkle_db(Arc::clone(&ledger_db), state_merkle_db)?;
+        }
+
+        Ok(())
+    }
+
+    fn get_current_version_in_ledger_db(ledger_db: &DB) -> Result<Option<Version>> {
+        let mut iter = ledger_db.iter::<TransactionInfoSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
+        Ok(iter.next().transpose()?.map(|item| item.0))
+    }
+
+    fn get_current_version_in_state_merkle_db(state_merkle_db: &DB) -> Result<Option<Version>> {
+        Self::find_closest_node_version_at_or_before(state_merkle_db, u64::max_value())
+    }
+
+    fn find_tree_root_at_or_before(
+        ledger_db: &DB,
+        state_merkle_db: &DB,
+        version: Version,
+    ) -> Result<Option<Version>> {
+        match Self::find_closest_node_version_at_or_before(state_merkle_db, version)? {
+            Some(closest_version) => {
+                if Self::root_exist_at_version(state_merkle_db, closest_version)? {
+                    return Ok(Some(closest_version));
+                }
+                let mut iter = ledger_db.iter::<EpochByVersionSchema>(ReadOptions::default())?;
+                iter.seek_for_prev(&version)?;
+                match iter.next().transpose()? {
+                    Some((closest_epoch_version, _)) => {
+                        if Self::root_exist_at_version(state_merkle_db, closest_epoch_version)? {
+                            Ok(Some(closest_epoch_version))
+                        } else {
+                            Ok(None)
+                        }
+                    },
+                    None => Ok(None),
+                }
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn root_exist_at_version(state_merkle_db: &DB, version: Version) -> Result<bool> {
+        Ok(state_merkle_db
+            .get::<JellyfishMerkleNodeSchema>(&NodeKey::new_empty_path(version))?
```

Comment:
try to see if I understand correctly. The key is (version, nibble_path). If we use empty_path, It will then return the root since its nibble_path is empty?

Reply:
Yes